processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,743 @@
+   * @param tablePrefix          Prefix for a join table
+   * @param clauseForTablePrefix Joinable clause for the prefix
+   * @param equiconditions       Map of equiconditions, keyed by the right hand columns
+   *
+   * @return A list of correlatation analyses for the equicondition RHS columns that reside in the table associated with
+   * the tablePrefix
+   */
+  @Nullable
+  private static List<JoinFilterColumnCorrelationAnalysis> findCorrelatedBaseTableColumns(
+      HashJoinSegmentStorageAdapter adapter,
+      String tablePrefix,
+      JoinableClause clauseForTablePrefix,
+      Map<String, Expr> equiconditions
+  )
+  {
+    JoinConditionAnalysis jca = clauseForTablePrefix.getCondition();
+
+    List<String> rhsColumns = new ArrayList<>();
+    for (Equality eq : jca.getEquiConditions()) {
+      rhsColumns.add(tablePrefix + eq.getRightColumn());
+    }
+
+    List<JoinFilterColumnCorrelationAnalysis> correlations = new ArrayList<>();
+
+    for (String rhsColumn : rhsColumns) {
+      List<String> correlatedBaseColumns = new ArrayList<>();
+      List<Expr> correlatedBaseExpressions = new ArrayList<>();
+      boolean terminate = false;
+
+      String findMappingFor = rhsColumn;
+      while (!terminate) {
+        Expr lhs = equiconditions.get(findMappingFor);
+        if (lhs == null) {
+          break;
+        }
+        String identifier = lhs.getBindingIfIdentifier();
+        if (identifier == null) {
+          // We push down if the function only requires base table columns
+          Expr.BindingDetails bindingDetails = lhs.analyzeInputs();
+          Set<String> requiredBindings = bindingDetails.getRequiredBindings();
+          for (String requiredBinding : requiredBindings) {
+            if (!adapter.isBaseColumn(requiredBinding)) {
+              return null;
+            }
+          }
+
+          terminate = true;
+          correlatedBaseExpressions.add(lhs);
+        } else {
+          // simple identifier, see if we can correlate it with a column on the base table
+          findMappingFor = identifier;
+          if (adapter.isBaseColumn(identifier)) {
+            terminate = true;
+            correlatedBaseColumns.add(findMappingFor);
+          }
+        }
+      }
+
+      if (correlatedBaseColumns.isEmpty() && correlatedBaseExpressions.isEmpty()) {
+        return null;
+      }
+
+      correlations.add(
+          new JoinFilterColumnCorrelationAnalysis(
+              rhsColumn,
+              correlatedBaseColumns,
+              correlatedBaseExpressions
+          )
+      );
+    }
+
+    return correlations;
+  }
+
+  private static boolean filterMatchesNull(Filter filter)
+  {
+    ValueMatcher valueMatcher = filter.makeMatcher(ALL_NULL_COLUMN_SELECTOR_FACTORY);
+    return valueMatcher.matches();
+  }
+
+  private static class AllNullColumnSelectorFactory implements ColumnSelectorFactory
```

Comment:
This class is pretty jam packed with different concepts and inner classes, maybe it would make sense to put it in its own `org.apache.druid.segment.join.filter` package and split it up into different classes? (Sort of like DataSourceAnalysis and its friend PreJoinableClause in `org.apache.druid.query.planning`, but this one is even more complex)

Reply:
I split out the inner classes and JoinFilterAnalyzer into a new package `org.apache.druid.segment.join.filter`